### PR TITLE
Feature: nakadi stream support

### PR DESCRIPTION
**Pull request by tkrop**
The test case for client exceptions in streams needs special REST client driver support that must be enabled by building from the branch https://github.com/tkrop/rest-driver/tree/feature/stream-and-lamda-support.

### CODE REVIEW

File: riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java [hunk 1]
```diff
@@ -84,6 +90,14 @@ public static RestConfigurer simpleRequestFactory(final ExecutorService executor
         };
     }
 
+    @Override
+    public RestBuilder clone() {
```
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java]: Why is this needed? Shouldn't the builder be re-usable by default?

File: riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java [hunk 2]
```diff
@@ -84,6 +90,14 @@ public static RestConfigurer simpleRequestFactory(final ExecutorService executor
...
+        return new RestBuilder()
+                .requestFactory(requestFactory)
+                .converters(converters)
+                .baseUrl(baseUrl);
+    }
+
     public Rest build() {
         return new Rest(requestFactory, converters(), baseUrl);
```
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java]: If we make a copy of the list here the builder can be re-used without cloning it.

File: riptide-core/src/main/java/org/zalando/riptide/Route.java [hunk 3]
```diff
@@ -22,28 +31,30 @@
...
+        };
     }
 
     static Route call(final ThrowingConsumer<ClientHttpResponse> consumer) {
-        return (response, reader) ->
+        return (response, reader) -> {
+            try {
                 consumer.accept(response);
+            } finally {
+                response.close();
```
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/Route.java]: If we somehow wrap this in a try-with-resource block we would get the additional support of surpressed exceptions, right?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Batch.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+import java.util.List;
+
+import lombok.Value;
+
+@Value
+final class Batch {
+
+    Cursor cursor;
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Batch.java]: Cursor should be of type JsonNode because we don't care about the actual content. Maybe a package private `@JsonValue JsonNode getValue()` to hide that from clients.

File: riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/RestAsyncClientHttpResponse.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        };
+    }
+
+    @Override
+    public HttpHeaders getHeaders() {
+        return response.getHeaders();
+    }
+
+    @Override
+    @SneakyThrows
```
Comment (jhorstmann) [inline riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/RestAsyncClientHttpResponse.java]: Sneaking around the code coverage requirements again :)

File: riptide-nakadi/pom.xml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        </dependency>
+
+        <!-- rest -->
+        <dependency>
+            <groupId>com.github.rest-driver</groupId>
+            <artifactId>rest-client-driver</artifactId>
+        </dependency>
+    </dependencies>
+ 
+    <distributionManagement>
```
Comment (jhorstmann) [inline riptide-nakadi/pom.xml]: This could be configured in the parent

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    private static final String SESSION_HEADER = "X-Nakadi-SessionId"; // TODO should this be X-Nakadi-StreamId?
+
+    private static final Binding<Series> ON_FAILURE_BINDING =
+            anySeries().call(dispatch(contentType(),
+                    on(PROBLEM).call(ThrowableProblem.class, propagate()),
+                    Bindings.anyContentType().call(invalidContent())));
+
+    private static ThrowingConsumer<ClientHttpResponse> invalidContent() {
+        return response -> {
+            InputStreamReader input = new InputStreamReader(response.getBody());
```
Comment (jhorstmann) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Should specify charset.

Instead of reading line by line and then joining you could also use guava `CharStream.toString(Readable)`.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    }
+
+    Metadata metadata;
+
+    String dataType;
+
+    DataOperation dataOp;
+
+    ObjectNode data;
```
Comment (jhorstmann) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: Do I see this correctly that you only support datachange events, and that users would need to databind from the `ObjectNode` to their own data model themselves?

File: riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java (hunk 1 above)
Comment (tkrop) [inline riptide-core/src/main/java/org/zalando/riptide/RestBuilder.java]: Only used at one location. Will remove it until more use cases are visible.

File: riptide-core/src/main/java/org/zalando/riptide/Route.java (hunk 3 above)
Comment (tkrop) [inline riptide-core/src/main/java/org/zalando/riptide/Route.java]: This would be in the anonymous dispatcher in the Requester.
```
future.addCallback(response -> {
                    try { ...
```
Doing it here would close the stream in any case. I think this would be okay for all use cases. Anyhow, I can't see how this would result in suppressed exceptions without having code coverage problems.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Batch.java (hunk 4 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Batch.java]: No, we care. We need to commit this cursor at the end.

Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Batch.java]: Yes, we need it, but we don't care about the actual content/properties.

File: riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/RestAsyncClientHttpResponse.java (hunk 5 above)
Comment (tkrop) [inline riptide-httpclient/src/main/java/org/zalando/riptide/httpclient/RestAsyncClientHttpResponse.java]: No, left over artifact of code change. Thanks for detecting it.

File: riptide-core/src/main/java/org/zalando/riptide/Route.java (hunk 3 above)
Comment (whiskeysierra) [inline riptide-core/src/main/java/org/zalando/riptide/Route.java]: Using try-with-resource will add any exception during close to the list of supressed exceptions (`Throwable.getSuppressed`) for further analysis (will be included in logfiles, iirc).

File: riptide-nakadi/pom.xml (hunk 6 above)
Comment (tkrop) [inline riptide-nakadi/pom.xml]: Yes, it is already there. I will remove this.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 7 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Although an option. Better is @whiskeysierra suggested:
```
new String(ByteStreams.toByteArray(response.getBody()), charset)
```
I like the above option, because it is without guava :)

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java (hunk 8 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: No, the support for datachange events should already be optional. The fields are available but must not be filled in by the stream.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 7 above)
Comment (jhorstmann) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Isn't `ByteStreams` also from guava?

File: riptide-core/src/main/java/org/zalando/riptide/Requester.java [hunk 9]
```diff
@@ -1,5 +1,7 @@
 package org.zalando.riptide;
 
+import static org.zalando.riptide.TryWith.tryWith;
```
Comment (lukasniemeier-zalando) [inline riptide-core/src/main/java/org/zalando/riptide/Requester.java]: ?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java (hunk 8 above)
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: I'd prefer to have `T` here (and the class to be generic) so clients don't have to do the mapping themselves.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import lombok.Value;
+
+@Value
+public class Event {
+
+    public enum DataOperation {
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: I'd make those inner classes top-level ones.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+import java.time.OffsetDateTime;
+import java.util.UUID;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import lombok.Value;
+
+@Value
+public class Event {
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: DataChangeEvent?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+        String flowId;
+
+    }
+
+    Metadata metadata;
+
+    String dataType;
+
+    DataOperation dataOp;
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]:
```java
@JsonProperty("data_op")
Operation operation;
```

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java (hunk 10 above)
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: Just Operation?!

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java (hunk 8 above)
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/Event.java]: > No, the support for datachange events should already be optional. The fields are available but need not to be filled in by the stream.

I wouldn't mix those different types of events in a single class.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            Charset charset = Optional.ofNullable(response.getHeaders().getContentType())
+                    .map(MediaType::getCharSet)
+                    .orElse((StandardCharsets.ISO_8859_1));
+            String body = new String(ByteStreams.toByteArray(response.getBody()), charset);
+            throw new IOException("Unsupported content: " + response.getRawStatusCode() +
+                    " - " + response.getStatusText() + "\n" + response.getHeaders() + "\n" + body);
+        };
+    }
+
+    private static ThrowingConsumer<ThrowableProblem> propagate() {
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Not needed, the propagate from the core will do the same thing.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        return entity -> {
+            throw entity;
+        };
+    }
+
+    static class GatewayException extends RuntimeException {
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: What's the use case for this? How does it help if it's not public?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+import org.zalando.riptide.Rest;
+import org.zalando.riptide.ThrowingConsumer;
+import org.zalando.riptide.capture.Capture;
+
+import com.google.common.io.ByteStreams;
+
+@Component
+public class NakadiGateway {
+
+    public static final MediaType PROBLEM = parseMediaType("application/problem+json");
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: We really should have a constant for that media type in the problem library.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(NakadiGateway.class);
+    private static final String SESSION_HEADER = "X-Nakadi-SessionId"; // TODO should this be X-Nakadi-StreamId?
+
+    private static final Binding<Series> ON_FAILURE_BINDING =
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: I'd use a Route here.

Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Doesn't have to be static, since the class is meant to be a singleton.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+import org.zalando.problem.ThrowableProblem;
+import org.zalando.riptide.Binding;
+import org.zalando.riptide.Bindings;
+import org.zalando.riptide.Rest;
+import org.zalando.riptide.ThrowingConsumer;
+import org.zalando.riptide.capture.Capture;
+
+import com.google.common.io.ByteStreams;
+
+@Component
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: I'd not use those Component/Autowired in here. Clients should wire them manually, since we can't have meaningful defaults.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    private final Rest rest;
+
+    @Autowired
+    public NakadiGateway(final Rest rest) {
+        this.rest = rest;
+    }
+
+    public void stream(final String owningApplication, final List<String> eventTypes, final EventConsumer consumer)
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Owning application should be a field/constructor argument, imho.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+            throws InterruptedException, ExecutionException, IOException {
+        stream(subscribe(new Subscription(null, owningApplication, eventTypes)), consumer);
+    }
+
+    private void stream(final Subscription subscription, final EventConsumer consumer)
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Since we're subscribing first and streaming later this method should be after `subscribe`.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 18 above)
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: I'd love to have this method generic, including the EventConsumer interface.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+
+    private ThrowingConsumer<ResponseEntity<Stream<Batch>>> process(final Subscription subscription,
+            final EventConsumer consumer) {
+        return entity -> {
+            @SuppressWarnings("resource")
+            final Stream<Batch> stream = entity.getBody();
+            try {
+                final HttpHeaders headers = entity.getHeaders();
+                stream.map(batch -> consume(consumer, batch))
+                        .filter(cursor -> cursor != null)
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]:
```java
.filter(Objects::nonNull)
```

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+    private Cursor consume(final EventConsumer consumer, final Batch batch) {
+        Cursor cursor = batch.getCursor();
+        if (cursor == null) {
+            LOG.debug("invalid batch: [{}]", batch);
+            return cursor;
+        }
+        List<Event> events = batch.getEvents();
+        if (events == null || events.isEmpty()) {
+            LOG.debug("empty batch: [{}]", batch);
+            return cursor;
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: I think returning null would be cleaner here, since there is nothing to commit, is there? @AlexanderYastrebov Thoughts, since you talked to the Nakadi guys about this?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+        }
+        events.forEach(event -> consumer.consume(event));
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]:
```java
events.forEach(consumer::consume);
```

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+        return cursor;
+    }
+
+    private Cursor[] commit(final Subscription subscription, final HttpHeaders headers, final Cursor... cursor) {
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Why not a `List<Cursor>`?

Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Why even having a return value at all? You're not using it anywhere, are you?

Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Instead of `HttpHeaders` I feel like a `String streamId` would be more honest in terms of what the method actually needs.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+            rest.put("/subscriptions/{id}/cursors", subscription.getId())
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .header(SESSION_HEADER, headers.getFirst(SESSION_HEADER))
+                    .body(cursor)
+                    .dispatch(series(),
+                            on(SUCCESSFUL).call(Cursor[].class, capture),
+                            ON_FAILURE_BINDING)
+                    .get();
+        } catch (ExecutionException ex) {
+            throw new GatewayException(ex.getCause());
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Why wrapping everything here?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(NakadiGateway.class);
+    private static final String SESSION_HEADER = "X-Nakadi-SessionId"; // TODO should this be X-Nakadi-StreamId?
+
+    private static final Binding<Series> ON_FAILURE_BINDING =
+            anySeries().call(dispatch(contentType(),
+                    on(PROBLEM).call(ThrowableProblem.class, propagate()),
+                    Bindings.anyContentType().call(invalidContent())));
+
+    private static ThrowingConsumer<ClientHttpResponse> invalidContent() {
```
Comment (whiskeysierra) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: `onUnknownErrorFormat`?

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 13 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: No, it will wrap it in an IOException, which I didn't like here to have.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 14 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Okay, it is not feature complete at the moment, and in a first approach I was not sure, whether I should unwrap this exception before giving the resulting exception to the consumer.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 16 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: There is currently nothing that hinders clients to build multiple NakadiGateways, and I see no reason to do so.

> I'd use a Route here.

It's still a binding, so I'm preferring it to name it that way.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 17 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Got me. Too much copy&past.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 18 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: > Owning application should be a field/constructor argument, imho.

Yes semantically agreed. And we have to include client_group_id here.

> I'd love to have this method generic, including the EventConsumer interface.

When we substitute the generic event against specific event types, this makes sense. As long as we do not plane to translate an generic event into a specific type, I would prefer to leave it that way.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 19 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: I prefer to have methods that continue each other to keep close together, especially if they have the same name. How about putting subscribe above streams, because you have to subscribe before streaming at all.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 21 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: True, didn't pushed this change yesterday.

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 23 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: > Why even having a return value at all? You're not using it anywhere, are you?

That's why I have not had a plan for this now. I even do not understand the reason why we have a list of coursers, while a batch contains only a single one...

File: riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java (hunk 25 above)
Comment (tkrop) [inline riptide-nakadi/src/main/java/org/zalando/riptide/nakadi/NakadiGateway.java]: Yes something in this direction would be good.

File: riptide-core/src/main/java/org/zalando/riptide/NoRouteException.java [hunk 26]
```diff
@@ -36,8 +43,26 @@ public NoRouteException(final ClientHttpResponse response) throws IOException {
...
-        return String.format("Unable to dispatch response (%d %s, Content-Type: %s)",
-                response.getRawStatusCode(), response.getStatusText(), response.getHeaders().getContentType());
+        return String.format("Unable to dispatch response: %d - %s\n%s\n%s",
+                response.getRawStatusCode(), response.getStatusText(), response.getHeaders(),
+                readStartOfBody(response, MAX_BODY_BYTES_TO_READ));
+    }
+
+    private static String readStartOfBody(final ClientHttpResponse response, final int length) throws IOException {
+        final InputStream stream = response.getBody();
+        if (stream != null) {
```
Comment (jhorstmann) [inline riptide-core/src/main/java/org/zalando/riptide/NoRouteException.java]: stream should be closed in a finally block

**Comment by AlexanderYastrebov** [comment]
Why is this made as a Riptide module? It is Nakadi-specific and should be separate library.
